modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -250,6 +321,35 @@ private void maybeExecuteRetention(ClusterState state, DataStream dataStream) {
                 );
             }
         }
+        return indicesToBeRemoved;
+    }
+
+    /*
+     * This method force merges any indices in the datastream that are not in the indicesToExclude set and that have not be force merged
+     * previously. Before force merging, it writes custom metadata as a marker in the cluster state so that we know a force merge is in
+     * process. And it writes another marker in the cluster state upon completion of the force merge.
+     */
+    private void maybeExecuteForceMerge(ClusterState state, DataStream dataStream, Set<Index> indicesToExclude) {
+        List<Index> readOnlyIndices = dataStream.getIndices().stream().filter(index -> indicesToExclude.contains(index) == false).toList();
+        Metadata metadata = state.metadata();
+        for (Index index : readOnlyIndices) {
+            if (dataStream.isIndexManagedByDLM(index, state.metadata()::index)) {
+                IndexMetadata backingIndex = metadata.index(index);
+                assert backingIndex != null : "the data stream backing indices must exist";
+                String indexName = index.getName();
+                boolean alreadyForceMerged = isForceMergeMarkerSet(backingIndex);
+                logger.trace("Already force merged {}: {}", indexName, alreadyForceMerged);
+                if (alreadyForceMerged == false) {
+                    ForceMergeRequest forceMergeRequest = new ForceMergeRequest(indexName);
+                    // time to force merge the index
+                    transportActionsDeduplicator.executeOnce(
+                        forceMergeRequest,
+                        new ErrorRecordingActionListener(indexName, errorStore),
+                        (req, reqListener) -> forceMergeIndex(forceMergeRequest, reqListener)
+                    );
```

Comment:
Oh good catch!

Reply:
I wound up adding them in an inner class in DataLifecycleService. The reason is that when I started adding them in ForceMergeRequest I saw that it has a field called `forceMergeUUID` that is always set to a random number in the constructor. It didn't seem right to have an equals/hashCode without `forceMergeUUID`, but it wouldn't work for deduplication. We discussed having a constructor that lets you pass in `forceMergeUUID`, but that gets written as user data to lucene commits so I was a little worried it would cause problems. So I wound up just doing it in a class whose only use is the `transportActionsDeduplicator` in DataLifecycleService.